### ISSUE TITLE
TLS connection is not started even if `LDAP_AUTH_USE_TLS = True`

### ISSUE THREAD

**Issue by githubuserx** [etianen/django-python3-ldap#247]
After recently having problems while logging in I traced the problem back to changes introduced in version **v0.15.0**.
While debugging it seems that the TLS connection is never started (at least in my environment) and the relevant code was removed in https://github.com/etianen/django-python3-ldap/commit/98b41250d4a925c55aa425a7ec5592e1b3352b4b#diff-27f3dc77469aa232297b4418ae9cbe0cb6fc9f58b5b5be8e32a017a323e0d78cL183-L185. The bind is then refused by the server.

I'm not sure if the idea was that the TLS connection should be started automatically, but after reading ldap3 documentation I *think* it is only started automatically if the correct `auto_bind` parameter is specified on the `Connection` object which doesn't apply in this case.

If I revert the changes mentioned above it all works again as expected. Maybe someone can confirm this. Thanks.

**Comment by etianen** [comment]
I think this was fixed in v0.15.2. Can you check?

**Comment by githubuserx** [comment]
I just checked again and the issue occured in version v0.15.2. In my previous post I was just referring to the version where the suspected changes were made.

The error from the server is:
```
LDAP bind failed: LDAPStrongerAuthRequiredResult - 8 - strongerAuthRequired - None - 00002028: LdapErr: DSID-0C090273, comment: The server requires binds to turn on integrity checking if SSL\TLS are not already active on the connection, data 0, v3839
```
<img width="1034" alt="screenshot" src="https://user-images.githubusercontent.com/5589968/181613921-84a1b77f-b96d-4986-affb-5dea7a2a9d22.png">

**Comment by EmilianoMorello** [comment]
Hi everyone, I have the same problem. With a different version it didn't happen.
"LDAP bind failed: LDAPStrongerAuthRequiredResult - 8 - strongerAuthRequired - None - 00002028: LdapErr: DSID-0C090259, comment: The server requires binds to turn on integrity checking if SSL\TLS are not already active on the connection, data 0, v4563  - bindResponse - None".

**Comment by etianen** [comment]
I have no idea what that error means, and no way of testing it. I'll take
an MR to fix it!

**Comment by githubuserx** [comment]
I've created a PR. @FlipperPA please comment if that change had some purpose that I'm not aware of.

**Comment by FlipperPA** [comment]
I’ll give this a test next week. It is odd, because the changes I made fixed my issue with the deprecated ciphers in Python 3.10; there was a definite before and after. Are we all connecting to an OpenLDAP server?

**Comment by githubuserx** [comment]
Ok, thanks. I'm also using Python 3.10 and never had any issues even with the earlier versions.
I don't know about OpenLDAP since in this case I'm connecting to a Windows server (domain controller) in Active Directory.

**Comment by EmilianoMorello** [comment]
> I’ll give this a test next week. It is odd, because the changes I made fixed my issue with the deprecated ciphers in Python 3.10; there was a definite before and after. Are we all connecting to an OpenLDAP server?

No, Microsoft LDAP.

**Comment by FlipperPA** [comment]
Aha, it is likely Microsoft and OpenLDAP have different behaviors. I may be able to get a test harness set up for both at work next week, I'll see.

To give you background on what inspired my PR, I was upgrading Python/Django running on Red Hat Enterprise 7.6 from Python 3.8 to Python 3.10. The Python `ssl` (well, TLS) module deprecated several TLS version ciphers, including `SSLv3` which `ldap3` was using under the hood:

https://docs.python.org/3/library/ssl.html#ssl.TLSVersion.TLSv1_3

The PR I issued allowed us to explicitly set the TLS Version cipher explicitly, and worked with OpenLDAP. My apologies if this has fouled up Active Directory. Hopefully, we can find a solution that works for all of us!

**Comment by githubuserx** [comment]
> The PR I issued allowed us to explicitly set the TLS Version cipher explicitly, and worked with OpenLDAP. 

I've seen that and as far as I can tell those changes are not the issue. The logic to start a TLS connection (please see #248, at least if `auto_bind=False`) on an otherwise plaintext socket seems independent of the stuff related to the TLS version and cipher. So I'm not sure if it was intentional to remove it. I assume there is a chance it only works now with OpenLDAP because it's not actually using TLS at all. Maybe worth checking that with Wireshark or server logs.

The thing that might be specific to MS AD is, that the server (when using recommended settings i.e. security baseline) rejects LDAP binds on a plaintext connection to prevent sending credentials unencrypted. This applies only to STARTTLS usually over port 389 (plaintext or TLS) and not for SSL over port 636.

**Comment by etianen** [comment]
What's the solution here? Do we do the revert in https://github.com/etianen/django-python3-ldap/pull/248? Or is there something better?

**Comment by githubuserx** [comment]
After some further research I can confirm that it is necessary to call `start_tls()` before any LDAP bind when the server requires LDAP signing. In general it would also be possible to start TLS after a bind, but not when using Active Directory.

This also corresponds to the ldap3 `auto_bind` parameters `AUTO_BIND_TLS_BEFORE_BIND` and `AUTO_BIND_TLS_AFTER_BIND`.
Since `auto_bind` is set to `False`  in this implementation, `start_tls()` must be called explicitly as it was the case up to version 0.14.0. Therefore I think this change is a clear regression and should be reverted as otherwise people might connect with an insecure connection and not notice it, depending on server configuration.

I also tested connecting with LDAPS over port 636 which always worked independent of this change, as expected.

- https://docs.microsoft.com/en-us/troubleshoot/windows-server/identity/enable-ldap-signing-in-windows-server#how-to-configure-the-directory-to-require-ldap-server-signing-for-ad-ds
- https://ldapwiki.com/wiki/LDAP%20Signing
- https://ldap3.readthedocs.io/en/latest/ssltls.html

An additional improvement would be to add settings for the other arguments of the ldap3 `Tls` object to allow for specifying certificates and validation options. This would be a separate issue though.

**Comment by etianen** [comment]
I've merged #248 , on the basis you sound like you've looked into this <3

**Comment by etianen** [comment]
Released as 0.15.3

**Comment by FlipperPA** [comment]
@etianen @githubuserx Sorry I'm just getting back to review this now - thanks for taking the bull by the horns.
